### DISCUSSION TITLE
Tutorial compared with stock git flow

### DISCUSSION THREAD

**Discussion by black7375** [arxanas/git-branchless#699] [Ideas]
First of all, thank you for making a great tool.
In my opinion, if there is a additional tutorial for the onboarding of new users, it will be helpful.

The current [Tutorial](https://github.com/arxanas/git-branchless/wiki/Tutorial) documentation explain key features well. :+1: 
However, users who are familiar with [stock git branch strategies](https://proandroiddev.com/get-git-flows-differences-under-your-belt-a31b7c96ebe7) seem to have difficulty switching or applying them right away. #402

I think it would be nice to explain the way that comes from a relatively universal `git flow` or `github flow` in the following way:
1. Description of [mental model through comparison](https://timothya.com/blog/git-stack/)
2. Examples that compare and practice step by step
   - Make one feature & reflect to main
   - Make multiple feature & reflect to main
   - Merge with conflict
   - Pull & push related work
3. Brief description of additional workflows ([Divergent development](https://github.com/arxanas/git-branchless/wiki/Workflow:-divergent-development), [Stashing](https://github.com/arxanas/git-branchless/wiki/Workflow:-stashing)..)

**Comment by arxanas** [comment]
That's a good idea! Unfortunately, I don't have time to work on the documentation right now, but I would be happy to accept or review documentation towards this end if anyone else wanted to write it.

**Reply by black7375** [reply]
I also don't have time right now, but I will try document work if possible by early next year.

**Comment by black7375** [comment]
Hello?
Recently, I have a chance to introduce this tool to team members who use GIT and briefly documented.
I intentionally used a lot of screenshots and I think it will help new users. #778

The purpose of this document focuses on how to maintain a linear branch(branchless) using the command of [Tutorial](https://github.com/arxanas/git-branchless/wiki/Tutorial).

If you like it, I hope it will be added to the wiki.
The current tutorial document is good as a summary.

---

## Learn step by step

### Init

First, we will initialize and create a general branch.
- Initialization with `git init` and `git branchless init`
- After `commit0` ~ `commit2`, make a branch called `featA` and proceed with `commit3`, `commit4`
- Returning to `main`

```shell
git init --initial-branch=main
git branchless init --main-branch=main

touch commit0
git add commit0
git commit -m "Add: commit0"

touch commit1
git add commit1
git commit -m "Add: commit1"

touch commit2
git add commit2
git commit -m "Add: commit2"

git switch -c featA
touch commit3
git add commit3
git commit -m "Add: commit3"

touch commit4
git add commit4
git commit -m "Add: commit4"

git switch main
```

As you can see from the log, it runs through `git hook`.
![image](https://user-images.githubusercontent.com/25581533/235311567-2c803fa1-504d-4f53-b75c-198e0bae63bd.png)
![image](https://user-images.githubusercontent.com/25581533/235311572-08aae010-c65d-40de-98ad-c35e9bcfa0a5.png)

Let's check the created commits by `git log`.
```shell
git log --graph --oneline --all --decorate --color --topo-order --branches -10 | tac
```
![image](https://user-images.githubusercontent.com/25581533/235311683-23afc0d3-b576-4840-8fc8-a50dbf828722.png)

Let's also check with `git sl`(`git smartlog`).
It makes information easier to see with simple command.
```shell
git sl
```
![image](https://user-images.githubusercontent.com/25581533/235311713-bdf6838f-ab81-436b-87bc-ef37d4432ba7.png)

### Basic to use

You will now create an anonymous branch.
You can see that they are both tracking well.
```shell
touch commit5
git add commit5
git commit -m "Add: commit5"

git switch --detach # Start anonymous branch

touch commit6
git add commit6
git commit -m "Add: commit6"
```
![image](https://user-images.githubusercontent.com/25581533/235311935-7ad48af2-89f1-4764-9fcf-586d6f87370a.png)
![image](https://user-images.githubusercontent.com/25581533/235311940-21930e2e-e529-4b90-9ab9-b921ea96567b.png)

Now let's say you want to go back to `main`, **the previous commit**. What should you do?
If it was stock git, it would have been called `git switch main`, but here is enough with `git prev`.
```
git prev
```
![image](https://user-images.githubusercontent.com/25581533/235312317-2434dafa-d611-4279-a5ad-71ae91e7c76c.png)

On the contrary, you can use `git next` to switch to the next commit.
If you put the number like `prev 3`, you switch to before 3 commits.
```shell
git next
git prev 3
```
![image](https://user-images.githubusercontent.com/25581533/235312479-f138fce1-8099-48fb-95e6-1333990b3989.png)
![image](https://user-images.githubusercontent.com/25581533/235312485-f13682a8-53b7-4025-9442-8d20ffa37ad7.png)

### Moving commits
Now, the `main` next is `commit6`.
But you might want to make `commit3` and `commit4` coming after `main`.
That's what it takes to keep a commit histoy linear.

Using `git sync` becomes a rebase based on `main`.
Let's switch to `commit3` and run it.
```shell
git next 2 -o # Go to the old commit direction, The new commit direction is `-n`
git sync
```
![image](https://user-images.githubusercontent.com/25581533/235313065-f842a67c-c8fe-4c1e-8f69-8b6778abda66.png)
![image](https://user-images.githubusercontent.com/25581533/235313068-6b64d9d7-c56d-4a1a-9144-f6c14ef1043a.png)

What if you regret after doing this?
It can be returned to the magic spell, `git undo`.
```shell
git undo
```
![image](https://user-images.githubusercontent.com/25581533/235313150-530d9865-f8d8-4dc0-91c2-d17aa3627d39.png)

Here comes one more advantage of `git sl`.
It is that it only tracks meaningful anonymous branches!!
In fact, I forgot the screenshot and turned it back to `git undo`, but it's dirty when you look at the original commit log graph.
![image](https://user-images.githubusercontent.com/25581533/235313380-b5dafc20-fe23-47a7-8434-66c921fb84a9.png)
(`commit3` and `commit4` are shown three times.)

This time, let's move `commit6` back to `commit4`.
```shell
git move -s 7de4f31 -d d4b489a
```

After `git undo`, you can also move the `featA` branch, `commit3` and `commit4`, to `commit6`.
```shell
git undo
git move -b featA -d 7de4f31
```
![image](https://user-images.githubusercontent.com/25581533/235313692-85578de4-caa4-4d81-9a35-52882e3c6de7.png)

### Keeping it branchless
With git-branchless, it is this easy to keep the commit graph without branch.
But when I thought about it, I wanted to rebase `commit6` and `commit3`.
```shell
git rebase -i HEAD~2
```
![image](https://user-images.githubusercontent.com/25581533/235313865-ffda4a4c-5e51-4e38-9d4c-9fe4f0cf4216.png)

A branch has been created because a new commit is created.
Fortunately, the `git restack` command in the guide will solve it neatly, so don't worry.
```shell
git restack
```
![image](https://user-images.githubusercontent.com/25581533/235314021-e6e0cbb0-ba94-4c92-a299-d1a387952428.png)

`git restack` can be applied not only to `git rebase` but also in `git commit --amend`.
If there is a conflict with `git sync` or `git move`, it [fails to explicitly](https://github.com/arxanas/git-branchless/wiki/Concepts#speculative-merges) add the `--merge` option, and you can always use `git undo` if you want to go back while merge.

Now enjoy a simpler and smarter commit graph, a [fearless git commands and commits](https://github.com/arxanas/git-branchless/wiki/Workflow:-divergent-development).

There is also a stronger and more intuitive [revset](https://github.com/arxanas/git-branchless/wiki/Reference:-Revsets) than [revision](https://git-scm.com/docs/revisions), so check if you need a selector.

**Reply by arxanas** [reply]
@black7375 Thank you very much! I linked to this post in the first paragraph of https://github.com/arxanas/git-branchless/wiki/Tutorial.

By the way, you can replace

```sh
echo "main" | git branchless init
```

with

```sh
git branchless init --main-branch=main
```

**Reply by black7375** [reply]
Thank you for advice 😄
